lib/types/index.d.ts

```diff
@@ -507,175 +547,169 @@ export namespace SourceCode {
-        TaggedTemplateExpression?: ((node: ESTree.TaggedTemplateExpression & NodeParentExtension) => void) | undefined;
-        "TaggedTemplateExpression:exit"?:
-        | ((node: ESTree.TaggedTemplateExpression & NodeParentExtension) => void)
-        | undefined;
-        TemplateElement?: ((node: ESTree.TemplateElement & NodeParentExtension) => void) | undefined;
-        "TemplateElement:exit"?: ((node: ESTree.TemplateElement & NodeParentExtension) => void) | undefined;
-        TemplateLiteral?: ((node: ESTree.TemplateLiteral & NodeParentExtension) => void) | undefined;
-        "TemplateLiteral:exit"?: ((node: ESTree.TemplateLiteral & NodeParentExtension) => void) | undefined;
-        ThisExpression?: ((node: ESTree.ThisExpression & NodeParentExtension) => void) | undefined;
-        "ThisExpression:exit"?: ((node: ESTree.ThisExpression & NodeParentExtension) => void) | undefined;
-        ThrowStatement?: ((node: ESTree.ThrowStatement & NodeParentExtension) => void) | undefined;
-        "ThrowStatement:exit"?: ((node: ESTree.ThrowStatement & NodeParentExtension) => void) | undefined;
-        TryStatement?: ((node: ESTree.TryStatement & NodeParentExtension) => void) | undefined;
-        "TryStatement:exit"?: ((node: ESTree.TryStatement & NodeParentExtension) => void) | undefined;
-        UnaryExpression?: ((node: ESTree.UnaryExpression & NodeParentExtension) => void) | undefined;
-        "UnaryExpression:exit"?: ((node: ESTree.UnaryExpression & NodeParentExtension) => void) | undefined;
-        UpdateExpression?: ((node: ESTree.UpdateExpression & NodeParentExtension) => void) | undefined;
-        "UpdateExpression:exit"?: ((node: ESTree.UpdateExpression & NodeParentExtension) => void) | undefined;
-        VariableDeclaration?: ((node: ESTree.VariableDeclaration & NodeParentExtension) => void) | undefined;
-        "VariableDeclaration:exit"?: ((node: ESTree.VariableDeclaration & NodeParentExtension) => void) | undefined;
-        VariableDeclarator?: ((node: ESTree.VariableDeclarator & NodeParentExtension) => void) | undefined;
-        "VariableDeclarator:exit"?: ((node: ESTree.VariableDeclarator & NodeParentExtension) => void) | undefined;
-        WhileStatement?: ((node: ESTree.WhileStatement & NodeParentExtension) => void) | undefined;
-        "WhileStatement:exit"?: ((node: ESTree.WhileStatement & NodeParentExtension) => void) | undefined;
-        WithStatement?: ((node: ESTree.WithStatement & NodeParentExtension) => void) | undefined;
-        "WithStatement:exit"?: ((node: ESTree.WithStatement & NodeParentExtension) => void) | undefined;
-        YieldExpression?: ((node: ESTree.YieldExpression & NodeParentExtension) => void) | undefined;
-        "YieldExpression:exit"?: ((node: ESTree.YieldExpression & NodeParentExtension) => void) | undefined;
+    interface NodeListener extends RuleVisitor {
+        ArrayExpression?(node: ESTree.ArrayExpression & NodeParentExtension): void;
```

Comment:
Changed these from optional properties that could be a function or `undefined` to just optional methods in order to properly extend `RuleVisitor`. I think this is functionally equivalent.